rascal-lsp/src/main/rascal/lang/rascal/tests/semanticTokenizer/Util.rsc

```diff
@@ -0,0 +1,175 @@
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+module lang::rascal::tests::semanticTokenizer::Util
+
+import IO;
+import List;
+import ParseTree;
+import String;
+
+// This is the main utility function to test the semantic tokenizer. All other
+// functions in this module are auxiliary (and private).
+bool testTokenizer(type[&T<:Tree] begin, str input, Expect expects...,
+        bool printActuals = false,
+        bool useSpecialCaseHighlighting = false,
+        bool applyRascalCategoryPatch = false) {
+
+    // First, compute the tokens by calling the semantic tokenizer (in Java)
+    Tree tree = parse(begin, input);
+    list[SemanticToken] tokens = toTokens(tree, useSpecialCaseHighlighting, applyRascalCategoryPatch);
+
+    // Next, compute the absolute location of each token (i.e., the position of
+    // each token in `tokens` is represented *relative to* its predecessor)
+    list[loc] locations = toAbsoluteLocations(input, tokens);
+
+    // Next, compute the string of each token
+    list[str] strings = [substring(input, l.offset, l.offset + l.length) | l <- locations];
+
+    // Last, test the expectations
+    list[Actual] actuals = sort(zip3(tokens, locations, strings), less);
+
+    if (printActuals) {
+        iprintln(actuals);
+    }
+
+    return (true | it && check(actuals, expect) | expect <- expects);
+}
+
+private list[loc] toAbsoluteLocations(str input, list[SemanticToken] tokens) {
+    list[str] lines = split("\n", input);
+
+    // Initialize the "cursor"
+    int line   = 0;
+    int column = 0;
+
+    return for (token <- tokens) {
+        column = token.deltaLine > 0 ? 0 : column; // Reset column?
+
+        // Advance the "cursor"
+        line   += token.deltaLine;
+        column += toLength32(token.deltaStart, substring(lines[line], column));
+
+        // Compute the absolute location of `token`
+        int offset = (0 | it + size(lines[i]) + 1 | i <- [0..line]) + column;
+        int length = toLength32(token.length, substring(lines[line], column));
+        append |unknown:///|(offset, length);
+    }
+}
+
+//
+// Semantic tokens
+//
+
+alias SemanticToken = tuple[
+    int deltaLine,
+    int deltaStart,
+    int length,
+    str tokenType,
+    str tokenModifier];
+
+@javaClass{org.rascalmpl.vscode.lsp.util.SemanticTokenizerTester}
+java list[SemanticToken] toTokens(Tree _, bool _, bool _);
+
+//
+// Length conversion: Characters are counted as 16-bit units in LSP, whereas
+// they are counted as 32-bit units in Rascal, so lengths need to be converted.
+// The difference manifests only in the presence of surrogate pairs.
+//
+
+alias Length32 = int;
+alias Length16 = int;
+
+private Length32 toLength32(Length16 n16, str input) {
```

Comment:
I'm very happy we have tests, but a bit sad we have to write all of this.

Reply:
Agreed...